### PR TITLE
Todo list sinatra app implementation

### PR DESCRIPTION
**Pull request by apunko**
Please review

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+source 'https://rubygems.org'
+
+ruby '2.4.0'
```
Comment (stahor) [inline Gemfile:3]: lock gems

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  end
+end
+
+use OmniAuth::Builder do
+  provider :github, secrets['github_key'], secrets['github_secret']
+end
+
+post '/add_item' do
+  if session[:authenticated]
+    item = Item.create(value:        params[:value],
```
Comment (stahor) [inline app.rb]: move arguments list to the newline
```
Item.create(
  value: params[:value],
  ...
)
```

File: app.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                       task_list_id: params[:task_list_id])
+    json item: { value: item.value, id: item.id }
```
Comment (stahor) [inline app.rb]: remove whitespaces from hash

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+delete '/items/:id' do
+  item = Item.find(params[:id])
+  item.destroy
+  json item_id: params[:id]
+end
+
+put '/items/:id' do
+  item = Item.find(params[:id])
+  item.update(done: !item.done)
+  json item: { done: item.done, id: item.id }
```
Comment (stahor) [inline app.rb]: and what if update have been failed for some reasons?

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+end
+
+get '/' do
+  @user = User.find(session[:user_id]) if session[:authenticated]
+
```
Comment (stahor) [inline app.rb]: newline?

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  erb :index
+end
+
+get '/auth/:provider/callback' do
+  auth = request.env['omniauth.auth']
+  @user = User.from_omniauth(auth)
+  if @user
+    session[:user_id] = @user.id
+    session[:user_name] = @user.name
```
Comment (stahor) [inline app.rb]: it's enough to store only user_id. Then you can implement current_user method that will fetch user from DB by id or return nil.

File: app.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+end
+
+put '/items/:id' do
+  item = Item.find(params[:id])
+  item.update(done: !item.done)
+  json item: { done: item.done, id: item.id }
+end
+
+get '/' do
+  @user = User.find(session[:user_id]) if session[:authenticated]
```
Comment (stahor) [inline app.rb]: replace with current_user check (look at the comment below)

File: app.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    erb "<h1>Can's create session</h1><h3>message:<h3> <pre>#{params}</pre>"
+  end
+end
+
+get '/auth/failure' do
+  erb "<h1>Authentication Failed:</h1><h3>message:<h3> <pre>#{params}</pre>"
+end
+
+get '/logout' do
+  session[:authenticated] = false
```
Comment (stahor) [inline app.rb]: u are not removing user_id and user_name from session here

File: config/database.yml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (stahor) [inline config/database.yml]: replace with example config. Your personal local config should not be under version control.

File: db/migrate/20170716231209_add_users.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class AddUsers < ActiveRecord::Migration[5.1]
+  def change
+    create_table :users do |t|
+      t.string :provider, null: false
+      t.string :uid, null: false
+    end
```
Comment (stahor) [inline db/migrate/20170716231209_add_users.rb:6]: always add timestamps to ur tables.

File: config/environments.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# The environment variable DATABASE_URL should be in the following format:
+# => postgres://{user}:{password}@{host}:{port}/path
+configure :production, :development do
```
Comment (stahor) [inline config/environments.rb]: Why do you need this file at all? I'm sure that there are much easier ways to connect to DB

File: models/task_list.rb [hunk 12]
```diff
@@ -0,0 +1,4 @@
+class TaskList < ActiveRecord::Base
+  belongs_to :user
+  has_many :items
```
Comment (stahor) [inline models/task_list.rb]: dependent: :destroy

File: models/user.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+class User < ActiveRecord::Base
+  has_many :task_lists
```
Comment (stahor) [inline models/user.rb]: dependent

File: models/user.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+  def self.from_omniauth(auth)
+    user = User.where(provider: auth.provider, uid: auth.uid).take
```
Comment (stahor) [inline models/user.rb]: use `find_by` instead

File: models/user.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    if user.nil?
```
Comment (stahor) [inline models/user.rb]: unless user

File: models/user.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+      user = User.create(provider: auth.provider,
```
Comment (stahor) [inline models/user.rb]: move arguments to the next line

File: models/user.rb [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+                         uid:      auth.uid,
+                         name:      auth.info.name)
+      TaskList.create(title: 'TODO list', user_id: user.id)
+    end
+
```
Comment (stahor) [inline models/user.rb]: remove newline

File: models/user.rb [hunk 18]
```diff
@@ -0,0 +1,15 @@
+class User < ActiveRecord::Base
+  has_many :task_lists
+
+  def self.from_omniauth(auth)
+    user = User.where(provider: auth.provider, uid: auth.uid).take
+    if user.nil?
+      user = User.create(provider: auth.provider,
+                         uid:      auth.uid,
+                         name:      auth.info.name)
+      TaskList.create(title: 'TODO list', user_id: user.id)
```
Comment (stahor) [inline models/user.rb]: `user.task_lists.create(title: 'TODO list')`

File: public/scripts/items_helper.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+  function createItemElement(item) {
+    var listElement = $("<li class='item' item_id='" + item.id + "''></li>");
+    listElement.append("<span class='check'>&#10003;</span>")
+      .append(item.value).append("<span class='close'>×</span>");
+    return listElement;
+  }
+
+  var items = {
+    Initialize: function(form_selector) {
```
Comment (stahor) [inline public/scripts/items_helper.js]: method names should be in camelCase in js code

File: public/scripts/items_helper.js [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+      _form_selector = form_selector;
+      $(form_selector).unbind('submit').submit(Forms.Items.OnCreate);
```
Comment (stahor) [inline public/scripts/items_helper.js]: unbind was deprecated. Use off instead

File: public/scripts/items_helper.js [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        e.preventDefault();
+        $.post({
+          url : "/add_item",
+          dataType: 'json',
+          data : $(_form_selector).serialize()
+        })
+        .done(function(e) {
+          var item_element = createItemElement(e.item);
+          $("ul.items").append(item_element);
+          $("ul.items").find("li:last .close").click(Forms.Items.OnDelete);
```
Comment (stahor) [inline public/scripts/items_helper.js]: `$("ul.items li:last .close")`

File: views/index.erb [hunk 22]
```diff
@@ -0,0 +1,5 @@
+<% if session[:authenticated] %>
+    <%= erb(:"task_lists/_view", :locals => { :task_list => @user.task_lists.first }) %>
```
Comment (stahor) [inline views/index.erb]: use new hash syntax

File: views/task_lists/_view.erb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+<% task_list = locals[:task_list] %>
+
+<div>
+  <div class="list_header">
+    <h2><%= task_list.title %></h2>
+    <%= erb(:"items/_new_form", :locals => { :task_list => task_list }) %>
```
Comment (stahor) [inline views/task_lists/_view.erb]: hash syntax

File: Gemfile (hunk 1 above)
Comment (apunko) [reply Gemfile:3]: Sorry, what do you mean? 
Do you mean bundle lock? bundle lock doesn't make anything for me.

File: db/migrate/20170716231209_add_users.rb (hunk 10 above)
Comment (apunko) [reply db/migrate/20170716231209_add_users.rb:6]: Added new migration.

**Comment by apunko** [comment]
Thanks for the review! can you check it again?

File: Gemfile (hunk 1 above)
Comment (stahor) [reply Gemfile:3]: lock file won't protect you from unexpected gem update if you type `bundle update sinatra`

File: db/schema.rb [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+  create_table "task_lists", force: :cascade do |t|
+    t.bigint "user_id"
+    t.string "title", null: false
+    t.index ["user_id"], name: "index_task_lists_on_user_id"
+  end
+
+  create_table "users", force: :cascade do |t|
+    t.string "provider", null: false
+    t.string "uid", null: false
+    t.string "name"
```
Comment (stahor) [inline db/schema.rb]: I  don't see timestamps) Run ur migrations

File: public/scripts/items_helper.js [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+      _form_selector = form_selector;
+      $(form_selector).off('submit').submit(Forms.Items.onCreate);
+      $("ul.items .close").click(Forms.Items.onDelete);
+      $("ul.items .check").click(Forms.Items.onUpdate);
+    },
+
+    onCreate: function (e) {
+        e.preventDefault();
+        $.post({
+          url : "/add_item",
```
Comment (stahor) [inline public/scripts/items_helper.js]: extra whitespaces detected

File: public/scripts/items_helper.js [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+          dataType: 'json',
+          data : $(_form_selector).serialize()
```
Comment (stahor) [inline public/scripts/items_helper.js]: the same

File: Gemfile [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+
+gem 'rubocop', '0.49.1', require: false
```
Comment (stahor) [inline Gemfile]: to development group

**Pull request merged**